components/inputs/input-date.js

```diff
@@ -297,19 +338,33 @@ class InputDate extends LocalizeCoreElement(LitElement) {
 		}
 	}
 
-	_handleSetToToday() {
+	async _handleSetToToday() {
 		const date = getToday();
-		this._updateValueDispatchEvent(formatDateInISO(date));
+		await this._updateValueDispatchEvent(formatDateInISO(date));
 		this._dropdown.close();
 		this.focus();
 	}
 
 	_setFormattedValue() {
-		this._formattedValue = this.value ? formatISODateInUserCalDescriptor(this.value) : (this.emptyText ? this.emptyText : '');
+		this._formattedValue = this._shownValue ? formatISODateInUserCalDescriptor(this._shownValue) : (this.emptyText ? this.emptyText : '');
 	}
 
-	_updateValueDispatchEvent(dateInISO) {
-		if (dateInISO === this.value) return;
+	async _updateValueDispatchEvent(dateInISO) {
+		if (dateInISO === this._shownValue) return; // prevent validation from happening multiple times for same change
+		this._shownValue = dateInISO;
+		if (dateInISO) {
+			if (this.minValue && getDateFromISODate(dateInISO).getTime() < getDateFromISODate(this.minValue).getTime()) {
+				this.setValidity({ rangeUnderflow: true });
+			} else if (this.maxValue && getDateFromISODate(dateInISO).getTime() > getDateFromISODate(this.maxValue).getTime()) {
+				this.setValidity({ rangeOverflow: true });
+			} else {
+				this.setValidity({});
+			}
+		} else {
+			this.setValidity({});
+		}
```

Comment:
Something that came to mind is since the validity flags are booleans you could rewrite this as:
```
this.setValidity({
    rangeUnderflow: dateInISO && this.minValue && getDateFromISODate(dateInISO).getTime() < getDateFromISODate(this.minValue).getTime(),
    rangeOverflow: dateInISO && this.maxValue && getDateFromISODate(dateInISO).getTime() > getDateFromISODate(this.maxValue).getTime()
});
```
Either should work, just depends what your preference is.

Reply:
Cool yeah that looks much cleaner - good idea!